src/library/scala/collection/immutable/ArraySeq.scala

```diff
@@ -208,18 +208,21 @@ object ArraySeq extends StrictOptimizedClassTagSeqFactory[ArraySeq] { self =>
 
   def empty[A : ClassTag]: ArraySeq[A] = emptyImpl
 
-  def from[A : ClassTag](it: scala.collection.IterableOnce[A]): ArraySeq[A] = unsafeWrapArray {
-    val n = it.knownSize
-    if (n > -1) {
-      val elements = Array.ofDim[A](n)
-      val iterator = it.iterator
-      var i = 0
-      while (i < n) {
-        ScalaRunTime.array_update(elements, i, iterator.next())
-        i = i + 1
-      }
-      elements
-    } else ArrayBuffer.from(it).toArray
+  def from[A](it: scala.collection.IterableOnce[A])(implicit tag: ClassTag[A]): ArraySeq[A] = it match {
+    case as: ArraySeq[A] if as.elemTag == tag => as
+    case _ => unsafeWrapArray {
+      val n = it.knownSize
+      if (n > -1) {
+        val elements = Array.ofDim[A](n)
```

Comment:
would `it.toArray` or `it.copyToArray` possibly be better? it ought to be more efficient for `s.c.m.ArraySeq` and `ArrayBuffer`.

Reply:
The code is currently almost the same as `Array.from` except for using an `ArrayBuffer` instead of `ArrayBuilder` when the size is unknown (which looks like a bad decision). So as a first improvement this whole default case could simply be `unsafeWrapArray(Array.from(it))`.

`toArray` in `IterableOnceOps` and `IterableOnceExtensionMethods` also use `ArrayBuffer`. This looks wrong. Why would you ever want to build an `Array[Object]` first and then (slow-)copy that to a potentially primitive array?

I suggest the following approach:
- In `IterableOnceExtensionMethods#toArray` always delegate to `toArray` for consistency.
- In `IterableOnceOps#toArray` use `ArrayBuilder` instead of `ArrayBuffer`.
- Implement `Array.from` via `toArray`. This looks odd and may be controversial. Usually companion objects are in charge of building a collection, but that's because they know best how to do it. `Array` is a basic type so *everyone* knows how to build it, but there's a lot of potential for building *from* other collections that are already array-based. It makes more sense to keep this knowledge in the source collection.
- And finally implement `ArraySeq.from` via `unsafeWrapArray(it.toArray)`.

Reply:
I'm a bit concerned about someone implementing `toArray` and not realising it's not safe to call `Array.from`